include/systems/diff_system.h

```diff
@@ -408,6 +411,21 @@ class DifferentiableSystem : public ImplicitSystem,
   void add_dot_var_dirichlet_bcs( unsigned int var_idx, unsigned int dot_var_idx);
 #endif
 
+private:
+  /**
+   * Stack of pointers to objects to use for physics assembly
+   * evaluations.  Physics assembly defaults to \p this for backwards
+   * compatibility if the stack is empty; for the most flexibility
+   * users should create separate physics objects.
+   */
+  std::vector<std::unique_ptr<DifferentiablePhysics>> _diff_physics;
+
+  /**
+   * Pointer to object to use for quantity of interest assembly
+   * evaluations.  Defaults to \p this for backwards compatibility; in
+   * the future users should create separate physics objects.
+   */
+  std::vector<std::unique_ptr<DifferentiableQoI>> _diff_qoi;
```

Comment:
These vectors could be `std::stacks` I guess... though its not clear to me if there are any benefits to that change beyond communicating their purpose to the reader. The runtime performance is probably exactly the same.